File: p.69/TS/solution.js
function primeFactors(n) {
    var factors = [];
    for (var i = 2; i < Math.sqrt(n); i++) {
        while (n % i === 0) {
            factors.push(i);
            n = Math.floor(n / i);
        }
    }
    if (n > 1) {
        factors.push(n);
    }
    return factors;
}
function totientFunction(factors) {
    var distinct_primes = factors.filter(function (value, index, self) { return self.indexOf(value) === index; });
    var result = 1;
    for (var _i = 0, distinct_primes_1 = distinct_primes; _i < distinct_primes_1.length; _i++) {
        var prime = distinct_primes_1[_i];
        result *= (1 - 1 / prime);
    }
    return 1 / result;
}
function main() {
    var value;
    var maxTotient = Number.NEGATIVE_INFINITY;
    for (var i = 2; i < 1000000; i++) {
        var factors = primeFactors(i);
        var totientValue = totientFunction(factors);
        if (totientValue > maxTotient) {
            maxTotient = totientValue;
            value = i;
        }
    }
    console.log("Number with maximum n/phi n is ".concat(value, " with ratio ").concat(maxTotient));
}
main();

File: p.69/Python/solution.py
from typing import List
from math import inf

def prime_factors(n: int) -> List[int]:
    factors = []
    for i in range(2, int(n**0.5) + 1):
        while n % i == 0:
            factors.append(i)
            n = n // i
    # in case n is itself is a prime number
    if n > 1:
        factors.append(n)
    return factors

# n/phi(n)
def totient_function(prime_factors: List[int]) -> float:
    result = 1
    for factor in set(prime_factors):
        result *= (1.0 - 1.0 / factor)
    return 1.0 / result

def main():
    number = None
    max_totient = -inf
    for i in range(2,1000_000):
        totient_value = totient_function(prime_factors(i))
        if totient_value > max_totient:
            max_totient = totient_value
            number = i
    print(number)

if __name__ == "__main__":
    main()
